### DISCUSSION TITLE
SmartDNS分流+v2raya配合使用

### DISCUSSION THREAD

**Discussion by jiange1236** [v2rayA/v2rayA#1411] [Ideas]
- 起因

v2ray单独分流的时候，下载大流量文件CPU占用高，易卡死，于是研究利用smartdns进行分流。

- 参考链接

https://github.com/felix-fly/v2ray-dnsmasq-dnscrypt?tab=readme-ov-file
https://github.com/pymumu/smartdns/issues/1238
https://pymumu.github.io/smartdns/config/forwarding-with-ipset/

- 具体实现

一、基本设置

1. 下载页面设置需自动下载的要分流的域名列表文件（也可上传）。(格式为一行一个域名，需要分流的域名)
2. 设置上游服务器，设置国外的DOT，DOH等服务器，并设置`服务器组`：比如oversea。可选勾选‘从默认组中排除’
   也可以设置一个上游服务器，通过代理走转发程序，设置`服务器组`用于专门分流查询，这样查询DNS的出口和数据出口一样，DNS结果的亲和度和更好。
3. 域名规则->域名分流设置：
   - 设置`域名分组`，名称为第二步的名称：比如'oversea'
   - 选择`跳过测速`
   - 选择`停用IPV6地址解析`
   - 选择`域名列表文件`，文件为第一步设置的自动下载的文件，此文件为需要分流的域名列表。
   - 对于临时个别要分流的域名，可以在`域名列表`文本框中输入。

二、最简单的按域名的透明代理：

如果smartdns运行在主路由上，那么可以使用REDIRECT模式对TCP进行透明代理。下面举例对TCP进行透明转发，其中的[ipsetname]是IP集合的名称，比如可以设置`proxy-list`：

1. shell命令，创建ipset集合
   `ipset create proxy-list hash:net timeout 600`
2. shell命令，iptable设置对应ipset集合中域名的TCP请求使用TPROXY重定向转发到本机2018的转发程序。
   `iptables -t nat -I PREROUTING -p tcp -m set --match-set proxy-list dst -j REDIRECT --to-ports 2018`
在luci的防火墙中同样记得添加一下ipset和iptables规则，不然重启会失效。
3. smartdns设置页面的`域名分流设置`->`ipset名称`设置为上述步骤中[ipsetname]的名称，比如`proxy-list`
4. 在主路由的2018端口，开启REDIRECT模式的转发程序。（即下面的第三个步骤）
5. 应用smartdns设置。
6. 结束。

三、配置v2raya：

附加配置目录:`/etc/v2raya/ext` 新增文件v2ray-ext.json
```
{"inbounds":[{"port":2018,"protocol":"dokodemo-door","listen":"0.0.0.0","sniffing":{"enabled":false,"metadataOnly":false},"settings":{"network":"tcp","followRedirect":true},"streamSettings":null,"tag":"redirect"}],"routing":{"rules":[{"type":"field","balancerTag":"proxy","inboundTag":["redirect"]}]}}
```

重启v2raya，搞定

**Comment by pk49800** [comment]
啊这，大佬有没有直接在GUI配置的做法？还有就是这个设置会影响原来在GUI上面的配置吗

**Reply by jiange1236** [reply]
只有最后一步 附加配置目录:/etc/v2raya/ext 新增文件v2ray-ext.json 是在luci里面操作，其他的都必须用终端执行。
